### PR TITLE
[BUG] Zerocoin: fix spend validation and remove obsolete sporks

### PR DESCRIPTION
**Pull request by random-zebra**
#2425 introduced two bugs which are preventing a successful chain sync from scratch on `master`:

1. It abstracted a piece of code in `ParseAndValidateZerocoinSpend`, which is only parsing/validating a single zerocoin spend per tx, while a transaction could contain multiple spends. Fix it by returning an optional list of `CoinSpendValue`, instead of a singleton.

2. It re-introduced in the validation the (unused-at-the-time) function `CheckPublicCoinSpendVersion`, which is now failing at the first block with (version 3) public spends.
As all spork-based checks, this was supposed to be skipped when `fInitialBlockDownload`. 
Zerocoin transactions are no longer accepted (in both nets) in the mempool, as well as inside blocks (after v5 enforcement), so we can just remove this obsolete function.

The third commit fixes a very old bug: a zerocoin spend tx `nValueIn` is counted twice in ConnectBlock (first directly, after the coinspend parsing/validation, and then via `CCoinsViewCache::GetValueIn`).

Then, since we are at it, we can deprecate `SPORK_16_ZEROCOIN_MAINTENANCE_MODE` and `SPORK_18_ZEROCOIN_PUBLICSPEND_V4`, which are now unused/unneeded.

The last three commits are performance optimizations to avoid multiple calls to `ContainsZerocoins()`/`HasZerocoinSpendInputs()` in `ConnectBlock`, when possible.

### CODE REVIEW

File: src/test/validation_tests.cpp [hunk 1]
```diff
@@ -94,25 +94,25 @@ BOOST_FIXTURE_TEST_CASE(test_simple_shielded_invalid, TestingSetup)
...
+void CheckBlockZcRejection(std::shared_ptr<CBlock>& pblock, int nHeight, CMutableTransaction& mtx, const std::string& expected_msg)
 {
     pblock->vtx.emplace_back(MakeTransactionRef(mtx));
     BOOST_CHECK(SolveBlock(pblock, nHeight));
     BlockStateCatcher stateCatcher(pblock->GetHash());
     stateCatcher.registerEvent();
     BOOST_CHECK(!ProcessNewBlock(pblock, nullptr));
     BOOST_CHECK(stateCatcher.found && !stateCatcher.state.IsValid());
-    BOOST_CHECK_EQUAL(stateCatcher.state.GetRejectReason(), "bad-blk-with-zc");
+    BOOST_CHECK_EQUAL(stateCatcher.state.GetRejectReason(), expected_msg);
```
Comment (furszy) [inline src/test/validation_tests.cpp:105]: now that we are here, would be good to add a check at the end for `tip != pblock->GetHash()`. Just to be extra sure that the block wasn't connected.

**Review by furszy — COMMENTED** [review]
Code review ACK f7a93cd4d4c1febee05e5074a1df25041d1aca83.
Going to try a fresh sync now.

File: src/test/validation_tests.cpp (hunk 1 above)
Comment (random-zebra) [reply src/test/validation_tests.cpp:105]: Added.

**Review by Fuzzbawls — APPROVED** [review]
ACK a5a20e8274e55d2bdc5e15a36ae154abf76d699f

Full sync on mainnet confirmed

**Review by furszy — APPROVED** [review]
ACK a5a20e8
Took a while to sync on a small VPS (and had a disk space abortion that ended up on a chain reindex..).

**Pull request merged**